src/com/xilinx/rapidwright/design/DesignTools.java

```diff
@@ -3926,22 +3940,102 @@ public static void prohibitPartialHalfSlices(Design design) {
                         bottomUsed = true;
                     }
                 }
+                if (c.getBEL().isFF()) {
+                    String belName = c.getBELName();
+                    char letter = belName.charAt(0);
+                    boolean isFF2 = belName.charAt(belName.length() - 1) == '2';
+                    String sitePinName = letter + (isFF2 ? "_I" : "X");
+                    Node n = si.getSite().getConnectedNode(sitePinName);
+                    if (used.contains(n)) {
+                        if (si.getCell(letter + "6LUT") == null && si.getCell(letter + "5LUT") == null) {
+                            // Add a 'user-routethru' cell to make input path available
+                            BELPin input = c.getBEL().getPin("D");
+                            Net net = si.getNetFromSiteWire(input.getSiteWireName());
+                            if (net == null) {
+                                EDIFHierCellInst inst = c.getEDIFHierCellInst();
+                                if (inst != null) {
+                                    EDIFHierPortInst portInst = inst.getPortInst("D");
+                                    if (portInst != null) {
+                                        EDIFHierNet logNet = portInst.getHierarchicalNet();
+                                        if (logNet != null) {
+                                            net = design.createNet(logNet);
+                                        }
+                                    }
+                                }
+                            }
+                            if (net != null) {
+                                SitePinInst spi = si.getSitePinInst(sitePinName);
+                                if (spi == null || (spi != null && !spi.getNet().equals(net))) {
+                                    BELPin lutInput = si.getBEL(letter + "6LUT").getPin("A6");
+                                    EDIFHierPortInst ffInput = c.getEDIFHierCellInst().getPortInst("D");
+                                    ECOTools.createAndPlaceInlineCellOnInputPin(design, ffInput, Unisim.LUT1,
+                                            si.getSite(), lutInput.getBEL(), "I0", "O");
+                                }
+                            }
+                        } else {
+                            throw new RuntimeException(
```

Comment:
Am I right in understanding that this branch is reached if `used.contains(n) && (si.getCell("6LUT") != null || si.getCell("5LUT") != null)`?

Is this sufficient? `used.contains(n)` simply means that it is used by any net -- that net could well be entering that flop anyway (and `6LUT`/`5LUT` is used as comb output).